provider/source/src/units/data.rs

```diff
@@ -32,70 +32,48 @@ impl DataProvider<UnitsDisplayNameV1Marker> for SourceDataProvider {
-        ) {
-            if let Some(unit) = unit {
-                map.insert(count, unit.to_string());
+        let mut patterns = BTreeMap::new();
+
+        let unit_patterns = match length {
+            "long" => &units_format_data.long,
+            "short" => &units_format_data.short,
+            "narrow" => &units_format_data.narrow,
+            _ => {
+                return Err(DataErrorKind::IdentifierNotFound
+                    .into_error()
+                    .with_debug_context(length))
             }
         }
-
-        fn populate_units_map(
-            unit_length_map: &BTreeMap<String, Patterns>,
-            unit: &str,
-        ) -> Result<BTreeMap<Count, String>, DataError> {
-            let mut result = BTreeMap::new();
-            // TODO(younies): this should be coming from the aux key or from the main key.
-            let legth_key = "length-".to_string() + unit;
-            let duration_key = "duration-".to_string() + unit;
-
-            let unit_length_map = match (
-                unit_length_map.get(&legth_key),
-                unit_length_map.get(&duration_key),
-            ) {
-                (Some(length), None) => length,
-                (None, Some(length)) => length,
-                _ => {
-                    return Err(DataErrorKind::IdentifierNotFound
-                        .into_error()
-                        .with_debug_context(unit))
-                }
-            };
-
-            add_unit_to_map_with_name(&mut result, Count::One, unit_length_map.one.as_deref());
-            add_unit_to_map_with_name(&mut result, Count::Two, unit_length_map.two.as_deref());
-            add_unit_to_map_with_name(&mut result, Count::Few, unit_length_map.few.as_deref());
-            add_unit_to_map_with_name(&mut result, Count::Many, unit_length_map.many.as_deref());
-            add_unit_to_map_with_name(&mut result, Count::Other, unit_length_map.other.as_deref());
-
-            Ok(result)
+        .iter()
+        // get the units which match the key after the `-` in the attribute
+        // For exmple,
+        //          if the unit is meter, the key `length-meter` will match.
+        //          if the unit is millisecond, the key `duration-millisecond` will match.
+        .find_map(|(key, patterns)| {
+            key.split_once('-')
+                .map(|(_, rest)| rest)
+                .filter(|&rest| rest == unit)
+                .map(|_| patterns)
+        })
+        .ok_or_else(|| DataErrorKind::InvalidRequest.into_error())?;
+
+        for (count, unit) in [
+            (Count::One, unit_patterns.one.as_deref()),
+            (Count::Two, unit_patterns.two.as_deref()),
+            (Count::Few, unit_patterns.few.as_deref()),
+            (Count::Many, unit_patterns.many.as_deref()),
+            (Count::Other, unit_patterns.other.as_deref()),
+        ] {
+            if let Some(unit) = unit {
+                patterns.insert(count, unit.to_string());
```

Comment:
nit: you can probably put `&str` in the `BTreeMap` instead of `String`.
```suggestion
                patterns.insert(count, unit);
```